sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1208,9 +1208,26 @@ object PushPredicateThroughJoin extends Rule[LogicalPlan] with PredicateHelper {
             reduceLeftOption(And).map(Filter(_, left)).getOrElse(left)
           val newRight = rightJoinConditions.
             reduceLeftOption(And).map(Filter(_, right)).getOrElse(right)
-          val newJoinCond = commonJoinCondition.reduceLeftOption(And)
-
-          Join(newLeft, newRight, joinType, newJoinCond)
+          val (newJoinConditions, others) =
+            commonJoinCondition.partition(canEvaluateWithinJoin)
+          val newJoinCond = newJoinConditions.reduceLeftOption(And)
+          // if condition expression is unevaluable, it will be removed from
+          // the new join conditions, if all conditions is unevaluable, we should
+          // change the join type to CrossJoin.
+          val newJoinType =
+            if (commonJoinCondition.nonEmpty && newJoinCond.isEmpty) {
```

Comment:
I think this should be done only in this case: https://github.com/apache/spark/pull/22326#discussion_r214813591

Reply:
Thanks for reminding, crossJoinEnable should be checked here.